File: src/cli/commands/assign.ts
/**
 * `agt assign` — Assign a todo to a member.
 */

import type { Command } from "commander";
import { findProject, readConfig } from "../../lib/project.js";
import { loadDoc, saveDoc } from "../../lib/storage.js";
import { updateTodo } from "../../lib/operations.js";
import {
  parseTodoRef,
  findTodoByNumber,
  findMember,
} from "../../lib/queries.js";
import { error, success } from "../output.js";

export function registerAssign(program: Command): void {
  program
    .command("assign")
    .description("Assign a todo to a member")
    .argument("<ref>", "Todo reference (e.g. ABC-1 or 1)")
    .argument("<member>", "Member name or ID")
    .option("--json", "Output as JSON")
    .action(
      async (ref: string, memberName: string, opts: { json?: boolean }) => {
        const paths = findProject();
        if (!paths) error("Not in an agt project. Run 'agt init' first.");

        const config = await readConfig(paths.configPath);
        let doc = await loadDoc(paths.dataPath);
        if (!doc) error("Project data not found.");

        const num = parseTodoRef(ref, config.prefix);
        if (num === null) error(`Invalid todo reference: "${ref}".`);

        const todo = findTodoByNumber(doc, num);
        if (!todo) error(`Todo ${config.prefix}-${num} not found.`);

        const member = findMember(doc, memberName);
        if (!member) error(`Member "${memberName}" not found.`);

        doc = updateTodo(doc, num, { assignee: member.id });
        await saveDoc(paths.dataPath, doc);

        const todoRef = `${config.prefix}-${num}`;
        if (opts.json) {
          console.log(
            JSON.stringify({
              ref: todoRef,
              assignee: member.name,
              assigneeId: member.id,
            }),
          );
        } else {
          success(`Assigned ${todoRef} to ${member.name}`);
        }
      },
    );
}

File: agt-rs/crates/agt-cli/src/commands/assign.rs
use anyhow::Result;

use agt_lib::operations::{self, UpdateTodoFields};
use agt_lib::queries;

use super::{load_project, parse_ref, save_project};

pub fn run(reference: String, member_name: String) -> Result<()> {
    let (paths, mut doc) = load_project()?;
    let (_, prefix, _, _) = queries::read_project_meta(&doc);
    let num = parse_ref(&reference, &prefix)?;

    let member = queries::find_member(&doc, &member_name)
        .ok_or_else(|| anyhow::anyhow!("Member \"{}\" not found", member_name))?;

    operations::update_todo(
        &mut doc,
        num,
        UpdateTodoFields {
            assignee: Some(Some(member.id.as_str())),
            ..Default::default()
        },
        None,
    )?;

    save_project(&paths, &mut doc)?;

    println!("Assigned {}-{} to {}", prefix, num, member.name);
    Ok(())
}
